File: cipher-ts/src/sbox.ts
// Key-derived S-box via Fisher-Yates — bifurcate/sbox.py.
import { ChaosEngine } from "./chaos.js";

export function generateSbox(engine: ChaosEngine): Uint8Array {
  const box = new Uint8Array(256);
  for (let i = 0; i < 256; i++) box[i] = i;
  for (let i = 255; i > 0; i--) {
    const j = engine.byte() % (i + 1);
    const t = box[i]!; box[i] = box[j]!; box[j] = t;
  }
  return box;
}

export function differenceRatio(a: Uint8Array, b: Uint8Array): number {
  let d = 0;
  for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) d++;
  return d / a.length;
}

File: bifurcate/sbox.py
"""Key-derived S-box via Fisher-Yates.

THE NOVELTY OF BIFURCATE lives here. Classical ciphers ship one fixed,
publicly known substitution table for every user on earth. Bifurcate grows the
table from the key, so two passphrases are effectively running two different
ciphers.

NO INVERSE TABLE IS NEEDED. In a Feistel network the round function F is never
inverted — the structure guarantees decryption regardless. That removes
an entire class of bug.
"""

from __future__ import annotations

from .chaos import ChaosEngine


def generate_sbox(engine: ChaosEngine) -> bytes:
    """Fisher-Yates shuffle of [0..255] driven by the chaotic stream.

    The shuffle is a permutation by construction, so the result is always a
    bijection — asserted in tests rather than assumed.
    """
    box = list(range(256))
    for i in range(255, 0, -1):
        j = engine.byte() % (i + 1)
        box[i], box[j] = box[j], box[i]
    return bytes(box)


def is_bijection(sbox: bytes) -> bool:
    """True iff every value 0..255 appears exactly once."""
    return len(sbox) == 256 and len(set(sbox)) == 256


def difference_ratio(a: bytes, b: bytes) -> float:
    """Fraction of positions where two S-boxes differ (key-sensitivity metric)."""
    if len(a) != len(b):
        raise ValueError("S-boxes must be the same length")
    return sum(1 for x, y in zip(a, b) if x != y) / len(a)
